engine/api/workflowtemplate/execute.go

```diff
@@ -38,85 +40,156 @@ func prepareParams(wt *sdk.WorkflowTemplate, r sdk.WorkflowTemplateRequest) inte
 	return m
 }
 
-func executeTemplate(t string, data map[string]interface{}) (string, error) {
-	tmpl, err := template.New(fmt.Sprintf("template")).Delims("[[", "]]").Parse(t)
+func parseTemplate(templateType string, number int, t string) (*template.Template, error) {
+	var id string
+	switch templateType {
+	case "workflow":
+		id = templateType
+	default:
+		id = fmt.Sprintf("%s.%d", templateType, number)
+	}
+
+	tmpl, err := template.New(id).Delims("[[", "]]").Parse(t)
 	if err != nil {
-		return "", sdk.WrapError(err, "cannot parse workflow template")
+		reg := regexp.MustCompile(`template: ([0-9a-zA-Z.]+):([0-9]+): (.*)$`)
+		submatch := reg.FindStringSubmatch(err.Error())
+		if len(submatch) != 4 {
+			return nil, sdk.WithStack(err)
+		}
+		line, err := strconv.Atoi(submatch[2])
+		if err != nil {
+			return nil, sdk.WithStack(err)
+		}
+		return nil, sdk.WithStack(sdk.WorkflowTemplateError{
+			Type:    templateType,
+			Number:  number,
+			Line:    line,
+			Message: submatch[3],
+		})
 	}
+	return tmpl, nil
+}
 
+func executeTemplate(tmpl *template.Template, data map[string]interface{}) (string, error) {
 	var buffer bytes.Buffer
 	if err := tmpl.Execute(&buffer, data); err != nil {
-		return "", sdk.WrapError(err, "cannot execute workflow template")
+		return "", sdk.WithStack(err)
 	}
-
 	return buffer.String(), nil
 }
 
+func decodeTemplateValue(value string) (string, error) {
+	v, err := base64.StdEncoding.DecodeString(value)
+	if err != nil {
+		return "", sdk.NewError(sdk.ErrWrongRequest, err)
+	}
+	return string(v), nil
+}
+
 // Execute returns yaml file from template.
-func Execute(wt *sdk.WorkflowTemplate, i *sdk.WorkflowTemplateInstance) (sdk.WorkflowTemplateResult, error) {
-	data := map[string]interface{}{
-		"id":     i.ID,
-		"name":   i.Request.WorkflowName,
-		"params": prepareParams(wt, i.Request),
+func Execute(wt *sdk.WorkflowTemplate, instance *sdk.WorkflowTemplateInstance) (sdk.WorkflowTemplateResult, error) {
+	result := sdk.WorkflowTemplateResult{
+		Pipelines:    make([]string, len(wt.Pipelines)),
+		Applications: make([]string, len(wt.Applications)),
+		Environments: make([]string, len(wt.Environments)),
 	}
 
-	v, err := base64.StdEncoding.DecodeString(wt.Value)
-	if err != nil {
-		return sdk.WorkflowTemplateResult{}, sdk.WrapError(err, "cannot parse workflow template")
+	var data map[string]interface{}
+	if instance != nil {
+		data = map[string]interface{}{
+			"id":     instance.ID,
+			"name":   instance.Request.WorkflowName,
+			"params": prepareParams(wt, instance.Request),
+		}
 	}
 
-	out, err := executeTemplate(string(v), data)
+	var parsingErrs []error
```

Comment:
we have a sdk.MultiError available, you have to check if it fits your need

Reply:
done